docs/development/text-handling.md

```diff
@@ -0,0 +1,278 @@
+containing information about its structure (like paragraphs, headings or lists).
+The information content of these sequences of words won't be of interest for the
+application, but the structure is required to derive formatting for various output
+contexts.**
+
+Although this might leave some gray areas, this definition could be used to decide
+if a given piece of information is or isn't text, e.g.:
+
+* **login names of users**: *no text*, because just one word (no spaces...) and no
+further structure possible. Also, the contained information is of interest for the
+application since we need it to derive which user just logged in.
+* **content of a mail**: *text* because certainly structured words are possible and
+common. The actual information content does matter for the receiving person, but not
+for the application.
+* **titles of objects**: *might or might not be text*, because we support a little
+structuring (bold and italics) currently, but it is debatable if we really want to
+do so in the future.
+
+
+## Requirements for Text Handling
+
+To provide ILIAS developers with a tool set to solve the outlined problems with the
+current `string`-based text handling approach, a new approach needs to implement
+the following requirements:
+
+* For a given piece of text it should be known at any time which markup is used
+in the underlying string representation of the text.
+* For a given piece of text it should be known at any time which structure elements
+could be used in the text.
+* On programmatic interfaces it should be possible to specify which structure and
+markup is required when passing text.
+* The tool set should support building user interfaces to input text with a specified
+markup and structure (but actually building said interfaces is out of scope here).
+* It should be possible to convert all texts to certain baseline representation.
+These are plain text (as this is a baseline that is supported in every interesting
+target context) and HTML (since this is the markup for browsers, the main environment
+of our users). These conversion can be lossful, though.
+* It may be possible to convert some text in a given representation to some other
+representation but in general it is only expected that every text can be converted
+to HTML and plain text.
+
+
+## Approach
+
+To solve the requirements we are looking to implement the following approach in
+a sub folder of `src\Data`, using the conventions and standards of that library.
+Conversions will also be made available via the `Refinery` to integrate them
+into an established framework.
+
+### Define Structure Options
+
+To make it possible to programmatically talk about structuring options for text,
+a central `class` (and later `enum`, once supported for PHP < 8.1 is cut) defines
+the options that we care about:
+
+```php
+class Structure
+{
+    const string BOLD = "bold";
+    const string ITALIC = "italic";
+    const string UNORDERED_LIST = "unordered_list";
```

Comment:
Yes, this would be a shared class. We would suggest additions via PR. Also, I would not want to target "everything that TinyMCE now does", but rather "a sensible list of text structures we could really support".

Also, I rather deliberatly tried to avoid the word "formatting" like hell (hope I suceeded) and talked about "structure" instead. AFAIK, many things that the Tiny does are about formatting, i.e. saying something like "make this 11px big", instead of giving structure, i.e. "this is a headline". I would target structure, not formatting. I see that "bold" and "italic" is questionable in that regard, though, but at least it is not about some defined font-weight, or something similar. We might find better names, though, and the line might be blurry in general.

The reason I would avoid talking about formatting is, that formatting mostly only makes sense in some context. Since we have a web app, this context can change easily, e.g. when switching from small to large display. So the formatting would be defined by some CSS (as it is supposed to be), while the text ist marked up with some structure information to do so.

Reply:
I very much like the approach to avoid anything resembling "formatting". The distinction might sometimes be hard, though, as you already point out in the case of bold and italic. It could be a way out to seek inspiration in the semantic elements defined in the ILIAS-Editor, thus avoiding anything with the slightest smell of formatting.